src/lib/utils/size_estimation_utils.hpp

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <random>
+
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+  * This function iterates over a given vector of strings and sums up the memory usage of all elements. Due to the
+  * small string optimization (SSO) in most current C++ libraries, each string has an initially allocated buffer (e.g.,
+  * 15 chars in GCC's libstdc++). If a string is larger, the string is allocated on the heap and the initial string
+  * object stores a pointer to the actual string on the heap.
+  */
+template <typename T>
+size_t memory_usage_of_string_objects(const T& string_vector) {
+  // Get the default capacity of SSO strings.
+  const auto sso_string_capacity = std::string("").capacity();
+
+  // Add size for pre-initialized strings. For compilers using SSO, sizeof(pmr_string) also includes the char buffer.
+  auto bytes = string_vector.capacity() * sizeof(pmr_string);
+
+  for (const auto& single_string : string_vector) {
+    if (single_string.length() > sso_string_capacity) {
+      // Length is guaranteed to return the number of bytes, independent from any potential encoding.
+      // For each string, \0 is appended to denote the end of the string.
+      bytes += single_string.length() + 1;
+    }
```

Comment:
No particular reason. Should be the same in this case.

Reply:
`capacity` is closer to the actually used size in memory. Thinking about it, you might even want to consider `malloc_usable_size`, depending on how you plan to use the data:

```
#include <iostream>
#include <cstdlib>
#include <string>
#include <malloc.h>

void malloc(

int main()
{
    auto s = std::string{"hallomartindieserstringistvolllang"};
    std::cout << s << "\n\tlength: " << s.length() << "\n\tcapacity: " << s.capacity() << "\n\tmalloc_usable_size: " << malloc_usable_size((void*)s.c_str()) << std::endl;
    s = "hallomartindieserstringistkuerzer";
    std::cout << s << "\n\tlength: " << s.length() << "\n\tcapacity: " << s.capacity() << "\n\tmalloc_usable_size: " << malloc_usable_size((void*)s.c_str()) << std::endl;
}
```

```
hallomartindieserstringistvolllang
	length: 34
	capacity: 34
	malloc_usable_size: 40
hallomartindieserstringistkuerzer
	length: 33
	capacity: 34
	malloc_usable_size: 40
```

Reply:
I am not sure this might be going "too far". Besides, I am not sure if the actual number if correct because `c_str()` (might?) creates a copy as it needs to add a null terminator for C-like strings to the end of the string. So the actual string object might be differently aligned than that new returned c_str.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/c_str

>The pointer obtained from c_str() may be invalidated by:
    Passing a non-const reference to the string to any standard library function, or
    Calling non-const member functions on the string, excluding operator[], at(), front(), back(), begin(), rbegin(), end() and rend(). 

Thus, you are working on the actual data.

Reply:
Hmm? I would read it the other way.